applications/swarm-nl.md

```diff
@@ -0,0 +1,156 @@
+
+As we started building the [SamaritanDB prototype](https://github.com/thewoodfish/samaritan-db) we realized that the networking layer will be a critical building block that can be abstracted into its own subsystem. This gave birth to the idea of creating SwarmNL: a networking layer that can be easily integrated into existing infrastructure without deep technical overhead and knowledge. As a standalone library, it will help teams building applications that require networking to move faster, reliably.
+
+Our goal is to build this for our own purposes and to make it available for any other project to use. Here's a diagram illustrating the integration of SwarmNL to SamaritanDB:
+
+![SwarmNL and SamaritanDB](https://raw.githubusercontent.com/thewoodfish/encode-ink/main/SwarmNL%20and%20SamDB.png)
+
+
+### Features
+
+After some research, here are the features we decided to focus on.
+
+* **Node Communication** -- SwarmNL enables P2P node communication by providing an interface for the following:
+    * **Node Configuration** -- SwarmNL would provide a simple interface to configure the node and specify parameters to dictate its behaviour. This includes:
+        * Selection and configuration of the transport layers to be supported by the node.
+        * Selection of the cryptographic keypair to use for identity generation e.g Edwards.
+        * Storage and retrieval of keypair locally.
+        * PeerID and multiaddress generation.
+        * Protocol specification and handlers.
+        * Event handlers for network events and logging.
+
+    * **Gossiping** -- SwarmNL will implement the [Gossipsub 1.1](https://github.com/libp2p/specs/blob/master/pubsub/gossipsub/gossipsub-v1.1.md) protocol, specified by the [libp2p spec](https://github.com/libp2p/specs). This would be easily configurable to accommodate varying and peculiar network conditions.
+        
+* **Scaling** -- SwarmNL needs to efficiently handle a growing (or shrinking) number of nodes while maintaining performance and reliability. Here's what we plan to implement to this effect:
+    * **Sharding** -- implementation of a flexible generic sharding protocol that allows application specify configurations like sharding hash functions and locations for shards.
+    * **Data Forwarding** -- definition of a protocol for forwarding messages between nodes in different shards and establishment of efficient routing mechanisms for inter-shard communication.
+    * **Fault Tolerance** -- implementation of fault-tolerant mechanisms for detecting (and recovering from) node failures. This might involve redundancy, node replication, erasure encoding/decoding or re-routing strategies. 
```

Comment:
Can you please elaborate on the fault tolerance you want to introduce that has a benefit over the already existing fault tolerance / error correction in the TCP / IP stack? I assume re-routing strategies are already handled by gossipsub but node replication sounds especially interesting.

Reply:
The fault tolerance mechanisms introduced in SwarmNL are designed to complement and extend the fault tolerance and error correction features provided by the traditional TCP/IP stack. While TCP/IP ensures reliable point-to-point communication, decentralized and peer-to-peer networks face unique challenges, such as dynamic node behaviors, varying network conditions, and potential node failures. SwarmNL enhances fault tolerance by introducing additional strategies beyond those typically found in the TCP/IP stack.
Replication is a key feature in SwarmNL's fault tolerance arsenal. Unlike traditional networks, where replication may not be as prevalent, SwarmNL allows nodes to serve as replicas, mirroring portions of the network. This replication approach enhances availability, as these replicas can seamlessly take over in case of a node failure. Replicated nodes may shadow specific network segments or act as backup nodes that consistently receive updates, ready to replace a failed node.
Also while gossipsub handles message dissemination, SwarmNL introduces additional message re-routing capabilities. In case of a node failure or unavailability, the system can intelligently reroute messages to specific nodes, ensuring that critical information reaches its destination without disruption.